### ISSUE TITLE
No basic addition and scalar multiplication features for MPS and MPO?

### ISSUE THREAD

**Issue by sujay-kazi** [ITensor/ITensors.jl#515]
It seems that straightfoward addition and scalar multiplication operations for MPSs and MPOs are not provided in the mps.jl and mpo.jl files. For instance, if I wanted to find the part of a state psi1 that is orthogonal to a different state psi2 and put that into a new state psi3, the natural thing I want to write down is
```
psi3 = psi1 - inner(psi1, psi2) * psi2
```
Or for instance, if I want to make sure a state satisfies a certain constraint, and I use a certain MPO G to identify and remove the "bad" components, then I may want to write something like
```
psi = psi - G * psi
```
But it seems that I cannot do either of these things. In particular, both attempting to multiply an MPS/MPO by a scalar and attempting to add together two MPSs or MPOs with matching structures produce errors, even though to me these seem like very natural things to do. I guess one could extract the ITensor object(s), perform the corresponding operation on that (those), and then make a new , but this seems needlessly roundabout.

Am I missing something? Is there a way to do this that I am just not noticing, or is there some reason not to have this functionality? Anyway, if I am indeed right that this functionality is not there, it would be great if this could be added to the package. If this is sufficiently straightforward, I would be willing to write it myself and then put a pull request, if you do not have time to do it.

**Comment by emstoudenmire** [comment]
Thanks, yes. It appears we are missing some operations. We do have `psi1+psi2` and `x*psi` where x is any scalar. But `psi*x` is missing so that is an obvious oversight. Also `psi1-psi2` is missing which we should have if we have multiplication and addition. So I'll check with @mtfishman who has been working more closely on these features, but I think he will be for including them.

On the other hand, I just want to note that we recommend a certain amount of caution when doing operations on MPS and MPO's using operator notation. One reason is that these operations can become expensive, whereas expressions like `psi1+a*psi2` don't always capture the "feel" of that. A more serious and concrete issue is that it can be important to pass truncation parameters, such as `maxdim`, to the algorithms used to perform arithmetic with MPS that requires one to use function notation. So like:
```
G = inner(psi1,psi2)
mGpsi2 = (-G)*psi2
psi3 = sum([psi1,mGpsi2];maxdim=100,cutoff=1E-4)
```

**Comment by sujay-kazi** [comment]
Thank you so much for the quick reply! If you don't mind me asking, what is the reason that maxdim and cutoff need to be used when just adding two MPSs to create a new one? I thought those were just for algorithms like DMRG.

**Comment by mtfishman** [comment]
Hi Sujay,

Users will usually want to set `maxdim` and `cutoff`, since adding two MPS exactly leads to an MPS with a bond dimension that is the sum of the bond dimensions of the two MPS. In general, it is best to try to do the operation approximately, which is what `maxdim` and `cutoff` would be used for. The goal would be for `+` to provide reasonable defaults for `maxdim` and `cutoff`, but it would err on the side of being conservative about the truncations being performed (so that answers don't end up too inaccurate), in which case users should generally set their own values for `cutoff` and `maxdim`, using the interface Miles mentioned, or the following:
```julia
add(psi1, psi2; maxdim = 100, cutoff = 1e-10)
+(psi1, psi2; maxdim = 100, cutoff = 1e-10)
```

Miles captured most of what I was going to say. Many of these operations are not included as an oversight. I was also not entirely happy with the implementation of scalar multiplication with an MPS/MPO using `*`, since I think right now it doesn't properly preserve the orthogonality of the MPS/MPO. There are also some slight subtleties about how you might choose to define `Number * MPS`. If there is a 1-site orthogonality center, the scalar should multiply just the orthogonality center, but if there isn't a well defined orthogonality center, one might want to distribute it evenly throughout the MPS in order to keep the norms of the individual tensors distributed evenly.

So in general, the operations you mention should be made available and were left out as an oversight. If you want to make a PR for them, please feel free.

-Matt

**Comment by emstoudenmire** [comment]
@mtfishman agreed there are some non-trivial things here. The point about spreading a scalar throughout an MPS or only multiplying by the orthogonality center is one good example. My initial thought is that we should only multiply into the orthogonality center if there is one, otherwise do the "spreading" idea by taking the Nth root of the scalar and multiplying it into all of the MPS tensors. But even for that idea, there are some non-trivial questions about efficiency (is it really so crucial that it calls for an O(N) algorithm versus O(1)?).

**Comment by mtfishman** [comment]
@emstoudenmire, good point about the cost being `O(N)` vs. `O(1)`. The case that I can think where it is important to spread the scalar evenly is normalizing a unitary MPO, where the normalization is `tr(U^T U) = 2^N`. In that case, it is useful to scale every site by something like `sqrt(2)` (for example), but what you scale by actually grows exponentially with the system size. So the cases where it is important to spread out the scale factor may require more subtle algorithms anyway (i.e. where you specify the scale per site, and it may be more appropriate to use a broadcast syntax like `U .*= sqrt(2)`). In general though, I agree it is probably ok to just scale a single tensor of the MPS/MPO (that are within the orthogonality bounds).

It is also important to keep in mind that these simple-looking algebraic manipulations between MPS/MPO can be very subtle. For example, doing something like:
```
psi1 - inner(psi1, psi2) * psi2
```
can cause numerical problems, since `inner(psi1, psi2)` can easily approach 0 or infinite as the system size `N` gets large. For that case, we have another function `loginner(psi1, psi2)` which outputs the logarithm of the inner product, and is computed in a way to avoid overflow/underflow problems. Using that function, the equivalent of the above expression could be something like:
```
psi1 - (loginner(psi1, psi2) / N) .* psi2
```
where `.*` multiplies each tensor of the MPS by the scale factor. This would need to handle the orthogonality in a nice way, however.

So TL;DR: algebraic manipulations between MPS/MPO may naively look simple, but each of them, even scalar multiplication, can have various subtleties.
